lib/3scale/backend/stats/cleaner.rb

```diff
@@ -0,0 +1,156 @@
+        # then, a script, periodic cron, etc. is responsible for calling
+        # delete!(). That method scans the whole database and deletes all the
+        # stats keys that belong to those services marked to be deleted.
+        # The downside of this method is that it requires direct access to the
+        # redis servers. Going through a proxy like Twemproxy does not work,
+        # because it does not support the "SCAN" command.
+        #
+        # In the past we tried an alternative approach. When we received a
+        # request to delete the stats of a service, we generated all the
+        # possible stats keys that could exist for it. That approach was not
+        # efficient because it ended up generating many keys that didn't exist
+        # and thus, unnecessary delete calls to Redis. That approach is also
+        # more complex and error prone. You can find the details here:
+        # https://github.com/3scale/apisonator/issues/90
+
+        include Storable
+
+        SET_WITH_SERVICES_MARKED_FOR_DELETION = 'set_with_services_marked_for_deletion'.freeze
+        private_constant :SET_WITH_SERVICES_MARKED_FOR_DELETION
+
+        SLEEP_BETWEEN_SCANS = 0.01 # In seconds
+        private_constant :SLEEP_BETWEEN_SCANS
+
+        SCAN_SLICE = 500
+        private_constant :SCAN_SLICE
+
+        STATS_KEY_PREFIX = 'stats/'.freeze
+        private_constant :STATS_KEY_PREFIX
+
+        class << self
+          include Logging
+          def mark_service_to_be_deleted(service_id)
+            storage.sadd(SET_WITH_SERVICES_MARKED_FOR_DELETION, service_id)
+          end
+
+          # Deletes all the stats for the services that have been marked for
+          # deletion.
+          #
+          # This method receives a collection of instantiated Redis clients.
+          # Those clients need to connect to Redis servers directly. They cannot
+          # connect to a proxy like Twemproxy. The reason is that this function
+          # needs to scan the database using the "SCAN" command, which is not
+          # supported by Twemproxy.
+          #
+          # The services marked as deletion will be marked as done only when
+          # this function finishes deleting the keys from all the Redis servers.
+          # This means that if the function raises in the middle of the
+          # execution, those services will be retried in the next call.
+          #
+          # Note 1: keys deleted cannot be restored.
+          # Note 2: this method can take a long time to finish as it needs to
+          # scan all the keys in several DBs.
+          def delete!(redis_conns, log_deleted_keys: false)
+            services = services_to_delete
+            logger.info("Going to delete the stats keys for these services: #{services.to_a}")
+
+            redis_conns.each do |redis_conn|
+              delete_keys(redis_conn, services, log_deleted_keys)
+            end
```

Comment:
We could run this in parallel for all the instances. Given how much it takes and all the sleeping, I think it's something we would want to have.

Reply:
I prefer to start with something simple. I'll leave that for a future PR. I suspect most users only have one shard anyway.